### Задание

In [ ]:
# Подключение класса для создания нейронной сети прямого распространения
from tensorflow.keras.models import Sequential
# Подключение класса для создания полносвязного слоя
from tensorflow.keras.layers import Dense
# Подключение оптимизатора
from tensorflow.keras.optimizers import Adam
# Подключение утилит для to_categorical
from tensorflow.keras import utils
# Подключение библиотеки для загрузки изображений
from tensorflow.keras.preprocessing import image
# Подключение библиотеки для работы с массивами
import numpy as np
# Подключение модуля для работы с файлами
import os
# Подключение библиотек для отрисовки изображений
import matplotlib.pyplot as plt
from PIL import Image
# Вывод изображения в ноутбуке, а не в консоли или файле
%matplotlib inline

In [ ]:
# Загрузка датасета из облака
import gdown
gdown.download('https://storage.yandexcloud.net/aiueducation/Content/base/l3/hw_pro.zip', None, quiet=True)

'hw_pro.zip'

In [ ]:
# Распаковываем архив hw_light.zip в папку hw_light
!unzip -q hw_pro.zip

In [ ]:
# Путь к директории с базой
base_dir = '/content/hw_pro'
# Создание пустого списка для загрузки изображений обучающей выборки
x_train = []
# Создание списка для меток классов
y_train = []
# Задание высоты и ширины загружаемых изображений
img_height = 20
img_width = 20
# Перебор папок в директории базы
for patch in os.listdir(base_dir):
    # Перебор файлов в папках
    for img in os.listdir(base_dir + '/' + patch):
        # Добавление в список изображений текущей картинки
        x_train.append(image.img_to_array(image.load_img(base_dir + '/' + patch + '/' + img,
                                                         target_size=(img_height, img_width),
                                                         color_mode='grayscale')))
        # Добавление в массив меток, соответствующих классам
        if patch == '0':
            y_train.append(0)
        else:
            y_train.append(1)
# Преобразование в numpy-массив загруженных изображений и меток классов
x_train = np.array(x_train)
y_train = np.array(y_train)
x_train = x_train.reshape(-1,400)
# Вывод размерностей
print('Размер массива x_train', x_train.shape)
print('Размер массива y_train', y_train.shape)


Размер массива x_train (102, 400)
Размер массива y_train (102,)


In [ ]:
model = Sequential()
model.add(Dense(100,input_dim=400,activation='relu'))
model.add(Dense(50,activation='relu'))
model.add(Dense(1,activation = 'sigmoid'))

model.compile(loss='binary_crossentropy',optimizer = 'adam',metrics = ['accuracy'])
utils.plot_model(model,to_file='model_plot.png',show_shapes=True)

model.fit(x_train,y_train,
          batch_size=35,
          epochs = 15,
          verbose = 1)




Epoch 1/15
3/3 [==============================] - 1s 7ms/step - loss: 130.6555 - accuracy: 0.5098
Epoch 2/15
3/3 [==============================] - 0s 6ms/step - loss: 46.2328 - accuracy: 0.5294
Epoch 3/15
3/3 [==============================] - 0s 5ms/step - loss: 30.4989 - accuracy: 0.5392
Epoch 4/15
3/3 [==============================] - 0s 6ms/step - loss: 15.9725 - accuracy: 0.7059
Epoch 5/15
3/3 [==============================] - 0s 8ms/step - loss: 21.7909 - accuracy: 0.5784
Epoch 6/15
3/3 [==============================] - 0s 6ms/step - loss: 13.6273 - accuracy: 0.6569
Epoch 7/15
3/3 [==============================] - 0s 6ms/step - loss: 9.0639 - accuracy: 0.7059
Epoch 8/15
3/3 [==============================] - 0s 7ms/step - loss: 6.1324 - accuracy: 0.7549
Epoch 9/15
3/3 [==============================] - 0s 6ms/step - loss: 5.0368 - accuracy: 0.8039
Epoch 10/15
3/3 [==============================] - 0s 6ms/step - loss: 2.1106 - accuracy: 0.8824
Epoch 11/15
3/3 [===============

In [ ]:
losss,accuracy = model.evaluate(x_test,y_test,verbose=0)
print(f'Тест loss {loss:.4f}')
print(f'Тест accuracy {accuracy:.4f}')



Добились на модели точности 0.9804 от 1 до 15 эпохи она возрастала, а функция ошибок уменьшалась с каждой эпохой со 130.655 до 0.2236
